### PR TITLE
Builds on .NET 4.7.2

### PR DESCRIPTION
**Pull request by unquietwiki**
#3755 was a dead-end: between the LINQ IEnumerable calls, and the changes in SSL/TLS support in 4.7.2, re-targeting to 4.7.2 solves the following issues...

- Compatible with VS 2022+
- Compatible with Windows 7,8,10,11 (and related Server builds)
- Most users should already have this or 4.8.x installed via Windows Update, or more recent OS inclusion.
- Resolves build issue demonstrated in https://stackoverflow.com/questions/62543481/tohashset-ienumerablestring
- Connects to remote servers correctly, unlike the "hacked" attempt to use 4.5 in VS 2022.
- .NET Framework 4.7.2 came out almost five years ago; 4.8.1 came out a few months ago, and had dropped support for the older OSes 4.7.2 was supporting.

Tests passed: 852 of 860; one of failures was an unpack test, but I could still use the app successfully to uninstall & install mods. The previous 4.5-oriented build didn't even get halfway through the tests.

I hope this balances the compatibility concerns for the user base and the ability to develop further on the software.

### CODE REVIEW

**Comment by unquietwiki** [comment]
Looking at the code with SonarLint, ModuleInstaller in Tests [has a missing assert](https://rules.sonarsource.com/csharp/RSPEC-1699) that matches the failed unpack test; so maybe that test was a false result?

**Comment by unquietwiki** [comment]
[ilrepack_successful_command.txt](https://github.com/KSP-CKAN/CKAN/files/10363369/ilrepack_successful_command.txt)

Okay, I dug into where ILRepack was having problems. This sequence works for me locally.

File: Tests/Tests.csproj [hunk 1]
```diff
@@ -42,15 +43,17 @@
...
+    <PackageReference Include="Moq" Version="4.17.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.2" />
+    <PackageReference Include="System.Linq" Version="4.3.0" />
     <PackageReference Include="YamlDotNet" Version="9.1.0" />
-    <PackageReference Include="NUnit" Version="3.12.0" />
+    <PackageReference Include="NUnit" Version="3.13.3" />
     <PackageReference Include="NUnit3TestAdapter" Version="3.16.1" />
   </ItemGroup>
-  <ItemGroup Condition=" '$(TargetFramework)' == 'net45' ">
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net462' ">
```
Comment (Olympic1) [inline Tests/Tests.csproj:56]: Is this meant to be `net462` or a typo?

File: Tests/Tests.csproj [hunk 2]
```diff
@@ -69,10 +72,15 @@
...
+    <ProjectReference Include="..\AutoUpdate\CKAN-autoupdate.csproj" />
+    <ProjectReference Include="..\Cmdline\CKAN-cmdline.csproj" />
+    <ProjectReference Include="..\ConsoleUI\CKAN-ConsoleUI.csproj" />
     <ProjectReference Include="..\Core\CKAN-core.csproj" />
+    <ProjectReference Include="..\GUI\CKAN-GUI.csproj" />
+    <ProjectReference Include="..\Netkan\CKAN-netkan.csproj" />
   </ItemGroup>
   <Choose>
-    <When Condition=" '$(TargetFramework)' == 'net45' ">
+    <When Condition=" '$(TargetFramework)' == 'net462' ">
```
Comment (Olympic1) [inline Tests/Tests.csproj:83]: Same as above

Comment (unquietwiki) [reply Tests/Tests.csproj]: Fixed: typo from original attempt at 462, now 472. Tests still at 852 of 860 & builds. Due apologies!

**Comment by HebaruSan** [comment]
Looks like this is the current build error.

```
ERROR: Failed to load assembly /__w/CKAN/CKAN/_build/out/CKAN-CmdLine/Debug/bin/net472/ICSharpCode.SharpZipLib.dll
Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
  at Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name, Mono.Cecil.ReaderParameters parameters) [0x000f7] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name) [0x00007] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.DefaultAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.MetadataResolver.Resolve (Mono.Cecil.TypeReference type) [0x00040] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Resolve (Mono.Cecil.TypeReference type) [0x00006] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.TypeReference.Resolve () [0x00006] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.Mixin.CheckedResolve (Mono.Cecil.TypeReference self) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeEnum (Mono.Cecil.TypeReference enum_type) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeElementValue (Mono.Cecil.TypeReference type) [0x0002f] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeElement (Mono.Cecil.TypeReference type) [0x00015] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeFixedArgument (Mono.Cecil.TypeReference type) [0x00015] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeConstructorArguments (Mono.Cecil.CustomAttribute attribute, Mono.Collections.Generic.Collection`1[T] parameters) [0x0002e] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.MetadataReader.ReadCustomAttributeSignature (Mono.Cecil.CustomAttribute attribute) [0x0003c] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.<Resolve>b__34_0 (Mono.Cecil.CustomAttribute attribute, Mono.Cecil.MetadataReader reader) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Read[TItem,TRet] (TItem item, System.Func`3[T1,T2,TResult] read) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.Resolve () [0x00017] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.get_ConstructorArguments () [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadCustomAttributes (Mono.Cecil.ICustomAttributeProvider provider) [0x0001b] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadModule (Mono.Cecil.ModuleDefinition module) [0x00085] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.<ReadModule>b__1_0 (Mono.Cecil.ModuleDefinition module, Mono.Cecil.MetadataReader reader) [0x00007] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Read[TItem,TRet] (TItem item, System.Func`3[T1,T2,TResult] read) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadModule () [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleReader.CreateModuleFrom (Mono.Cecil.PE.Image image, Mono.Cecil.ReaderParameters parameters) [0x0006a] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.IO.Stream stream, Mono.Cecil.ReaderParameters parameters) [0x00028] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x0000a] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.AssemblyDefinition.ReadAssembly (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.ReadInputAssembly (System.String assembly, System.Boolean isPrimary) [0x00151] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.ReadInputAssemblies () [0x00048] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.Repack () [0x0002f] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.Application.Main (System.String[] args) [0x0004a] in <fa257496929944d3922e0349a0420534>:0 
An error occurred when executing task 'Repack-Ckan'.
```

**Comment by unquietwiki** [comment]
@HebaruSan taking a look, thanks for the heads up. I told it to pull in all the new commits, so I'll see how that impacts things, and will report back.

**Comment by unquietwiki** [comment]
@HebaruSan I found the missing NuGet include, and my latest build does run for me. Sneaky because it was building in VS just fine, though still missing that dependency.

**Comment by HebaruSan** [comment]
Now it says this:

```
ERROR: Failed to load assembly /__w/CKAN/CKAN/_build/out/CKAN-CmdLine/Release/bin/net472/ICSharpCode.SharpZipLib.dll
Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
  at Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name, Mono.Cecil.ReaderParameters parameters) [0x000f7] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.BaseAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name) [0x00007] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.DefaultAssemblyResolver.Resolve (Mono.Cecil.AssemblyNameReference name) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.MetadataResolver.Resolve (Mono.Cecil.TypeReference type) [0x00040] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Resolve (Mono.Cecil.TypeReference type) [0x00006] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.TypeReference.Resolve () [0x00006] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.Mixin.CheckedResolve (Mono.Cecil.TypeReference self) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeEnum (Mono.Cecil.TypeReference enum_type) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeElementValue (Mono.Cecil.TypeReference type) [0x0002f] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeElement (Mono.Cecil.TypeReference type) [0x00015] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeFixedArgument (Mono.Cecil.TypeReference type) [0x00015] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.SignatureReader.ReadCustomAttributeConstructorArguments (Mono.Cecil.CustomAttribute attribute, Mono.Collections.Generic.Collection`1[T] parameters) [0x0002e] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.MetadataReader.ReadCustomAttributeSignature (Mono.Cecil.CustomAttribute attribute) [0x0003c] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.<Resolve>b__34_0 (Mono.Cecil.CustomAttribute attribute, Mono.Cecil.MetadataReader reader) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Read[TItem,TRet] (TItem item, System.Func`3[T1,T2,TResult] read) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.Resolve () [0x00017] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.CustomAttribute.get_ConstructorArguments () [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadCustomAttributes (Mono.Cecil.ICustomAttributeProvider provider) [0x0001b] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadModule (Mono.Cecil.ModuleDefinition module) [0x00085] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.<ReadModule>b__1_0 (Mono.Cecil.ModuleDefinition module, Mono.Cecil.MetadataReader reader) [0x00007] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.Read[TItem,TRet] (TItem item, System.Func`3[T1,T2,TResult] read) [0x00025] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ImmediateModuleReader.ReadModule () [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleReader.CreateModuleFrom (Mono.Cecil.PE.Image image, Mono.Cecil.ReaderParameters parameters) [0x0006a] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.IO.Stream stream, Mono.Cecil.ReaderParameters parameters) [0x00028] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.ModuleDefinition.ReadModule (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x0000a] in <fa257496929944d3922e0349a0420534>:0 
  at Mono.Cecil.AssemblyDefinition.ReadAssembly (System.String fileName, Mono.Cecil.ReaderParameters parameters) [0x00000] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.ReadInputAssembly (System.String assembly, System.Boolean isPrimary) [0x00151] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.ReadInputAssemblies () [0x00048] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.ILRepack.Repack () [0x0002f] in <fa257496929944d3922e0349a0420534>:0 
  at ILRepacking.Application.Main (System.String[] args) [0x0004a] in <fa257496929944d3922e0349a0420534>:0 
An error occurred when executing task 'Repack-Ckan'.
```

FYI, you can always see the latest errors on the Checks tab, or by clicking Details in the box at the bottom of the PR page.

**Comment by unquietwiki** [comment]
@HebaruSan I looked at the include again: the newer version of SharpZipLib was asking for .NET Standard 2.0; which from what I know in the Unity/Kerbal space, there's a choice between .NET 4.x and Standard 2.x for builds, and this build setup is meant for the former. It was probably working just fine for VS builds, but not for Mono builds. I've reverted the version of the library to try to avoid that dependency.

**Comment by unquietwiki** [comment]
@HebaruSan Okay, that's a better outcome. There's some Docker timeout on two of the builds, and the remaining two seem related to whatever you were using .Net 5 for (I don't fully understand how you were able to make that work; it seems only Core & Tests use it). The rest worked!

**Comment by HebaruSan** [comment]
Cool, yeah, that's progress. 🎉 

We do need all the tests to pass before merging, though, and I don't have the spare cognitive bandwidth to dig into this. Maybe try reverting parts till you find out what breaks it?

**Comment by unquietwiki** [comment]
@HebaruSan I'll see what I can figure out. 👍  If you can tell me what you were trying to do with .NET 5, as you can remember, that'd be a good help. One other thing regarding that: .NET 6 is the current LTS, so whatever you're doing with 5 should probably move to that.

**Comment by HebaruSan** [comment]
I think you can find most of it in [the git log for the core csproj](https://github.com/KSP-CKAN/CKAN/commits/master/Core/CKAN-core.csproj). Originally @jbrot submitted .NET Standard 2.0 support in #2820 with plans to create a new Mac GUI that never materialized, then later @DasSkelett updated that to .NET 5 in #3194 in anticipation of a future MAUI GUI, before Microsoft announced that would leave Linux and Mac users out in the cold.

So it's not in use, but it creates opportunities to migrate to more modern platforms.

**Comment by unquietwiki** [comment]
Thanks for that; that's helpful! When I first started looking at this, I was trying to follow some of what's [suggested by Microsoft](https://learn.microsoft.com/en-us/dotnet/core/porting/?source=recommendations), regarding migration to .NET 5-7. I can try to figure out how to at least fix the build issue; otherwise, a separate branch using the other migration tools might yield the desired upgrade.

**Comment by unquietwiki** [comment]
@HebaruSan I noticed in the logic there were calls to two different file managers, depending on .NET5 vs 4.x. I took care of trying to fix that reference, tried calling on .NET6, and disabled a 4.5-specific Hashset method (LINQ seems to provide it). Still don't have an easy way to test on my end: VS doesn't like it that most of the other projects don't target the newer framework, even with added references to try that.

**Comment by HebaruSan** [comment]
> Still don't have an easy way to test on my end

I think you can run this from the command line:

```
.\build.ps1 test --configuration=Debug_NetCore
```

**Comment by HebaruSan** [comment]
And yes, `ToHashSet` was added to .NET Framework in 4.7.2, after CKAN had been using its own implementation of the same function for several years. That's why it has that `#if NET45` check.

https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.tohashset?view=netframework-4.7.2

Product | Versions
-- | --
.NET | Core 2.0, Core 2.1, Core 2.2, Core 3.0, Core 3.1, 5, 6, 7
.NET Framework | 4.7.2, 4.8, 4.8.1
.NET Standard | 2.1

**Comment by unquietwiki** [comment]
@HebaruSan that helped, thanks. I merged the last few commits that were coming up to the .NET5 build issue; and also gave up on using .NET6, since the builder on here is still on .NET5 (and the local compiler warned me plenty .NET5 is EOL).

Regarding the incompatibility warnings I keep seeing... the messages I saw involved two of the libraries being potentially incompatible with .NET 5. One's CommandLineParser: I tried creating separate references between the current one and a .NET Standard version for .NET 5 to use; that just confused VS; going to version 2.x on that library requires a partial re-write of the code here. The other was iniparser: having the .NET Standard version alongside isn't throwing any immediate errors.

**Comment by HebaruSan** [comment]
@unquietwiki, if you come back to this, please let us know and we can re-open it. Closing to reduce inactive stuff accumulating in the queue.

**Comment by unquietwiki** [comment]
@HebaruSan hey there, checking in; sorry for the delay. I'd like to finish this, but I guess I need some help from someone that worked on the original .NET 5 hooks & if they're even needed anymore. I also feel like since .NET 5 is EOL, the tests should be updated to .NET 6 LTS; that's a call you'd have to make. Let me know your thoughts. Thanks.

**Comment by HebaruSan** [comment]
I don't think you're going to get that help; as sometimes happens, you're the one that's interested in this, so whatever you can figure out is what will get done. Good luck!

**Comment by unquietwiki** [comment]
Thanks @HebaruSan. BTW, somewhat related to this: I saw a PR #3131 that updates CommandLineParser. That probably needs some love to ensure it still works, but that is a dependency issue I ran into on my PR.

**Pull request closed without merging**